File: application/monorepo_for_all_services/code/src/BoundedContext/CreditCardProduct/Product/Aggregate/Product.php
<?php

declare(strict_types=1);

namespace Galeas\Api\BoundedContext\CreditCardProduct\Product\Aggregate;

use Galeas\Api\BoundedContext\CreditCardProduct\Product\ValueObject\PaymentCycle;
use Galeas\Api\BoundedContext\CreditCardProduct\Product\ValueObject\Reward;
use Galeas\Api\Common\Aggregate\Aggregate;
use Galeas\Api\Common\Aggregate\AggregateTrait;
use Galeas\Api\Common\Id\Id;

class Product implements Aggregate
{
    use AggregateTrait;

    private string $name;
    private int $interestInBasisPoints;
    private int $annualFeeInCents;
    private PaymentCycle $paymentCycle;
    private int $creditLimitInCents;
    private int $maxBalanceTransferAllowedInCents;
    private Reward $reward;

    private string $cardBackgroundHex;

    private bool $isActive;

    public static function fromProperties(
        Id $aggregateId,
        int $aggregateVersion,
        string $name,
        int $interestInBasisPoints, 
        int $annualFeeInCents, 
        PaymentCycle $paymentCycle, 
        int $creditLimitInCents, 
        int $maxBalanceTransferAllowedInCents,
        Reward $reward,
        string $cardBackgroundHex,
        bool $isActive
    ): self {
        $product = new self($aggregateId, $aggregateVersion);
        $product->name = $name;
        $product->interestInBasisPoints = $interestInBasisPoints;
        $product->annualFeeInCents = $annualFeeInCents;
        $product->paymentCycle = $paymentCycle;
        $product->creditLimitInCents = $creditLimitInCents;
        $product->maxBalanceTransferAllowedInCents = $maxBalanceTransferAllowedInCents;
        $product->reward = $reward;
        $product->cardBackgroundHex = $cardBackgroundHex;
        $product->isActive = $isActive;

        return $product;
    }


    public function name(): string
    {
        return $this->name;
    }


    public function interestInBasisPoints(): int
    {
        return $this->interestInBasisPoints;
    }


    public function annualFeeInCents(): int
    {
        return $this->annualFeeInCents;
    }

    public function paymentCycle(): PaymentCycle
    {
        return $this->paymentCycle;
    }


    public function creditLimitInCents(): int
    {
        return $this->creditLimitInCents;
    }

    public function maxBalanceTransferAllowedInCents(): int
    {
        return $this->maxBalanceTransferAllowedInCents;
    }

    public function reward(): Reward
    {
        return $this->reward;
    }


    public function cardBackgroundHex(): string
    {
        return $this->cardBackgroundHex;
    }


    public function isActive(): bool
    {
        return $this->isActive;
    }
}

File: application/backend-credit-card-enrollment/backend-c#/CreditCardEnrollment/Domain/Product/Product.cs
using CreditCardEnrollment.Common.Aggregate;
using CreditCardEnrollment.Common.Events;
using CreditCardEnrollment.Domain.Product.Events;

namespace CreditCardEnrollment.Domain.Product;

public class Product : Aggregate
{
    public string Name { get; private set; } = string.Empty;
    public int InterestInBasisPoints { get; private set; }
    public int AnnualFeeInCents { get; private set; }
    public string PaymentCycle { get; private set; } = string.Empty;
    public int CreditLimitInCents { get; private set; }
    public int MaxBalanceTransferAllowedInCents { get; private set; }
    public string Reward { get; private set; } = string.Empty;
    public string CardBackgroundHex { get; private set; } = string.Empty;
    public bool IsActive { get; private set; }

    public static Product Define(
        string eventId,
        string aggregateId,
        string correlationId,
        string name,
        int interestInBasisPoints,
        int annualFeeInCents,
        string paymentCycle,
        int creditLimitInCents,
        int maxBalanceTransferAllowedInCents,
        string reward,
        string cardBackgroundHex)
    {
        var product = new Product();
        
        var @event = new ProductDefined
        {
            EventId = eventId,
            AggregateId = aggregateId,
            AggregateVersion = 1,
            CorrelationId = correlationId,
            CausationId = eventId,
            RecordedOn = DateTime.UtcNow,
            Name = name,
            InterestInBasisPoints = interestInBasisPoints,
            AnnualFeeInCents = annualFeeInCents,
            PaymentCycle = paymentCycle,
            CreditLimitInCents = creditLimitInCents,
            MaxBalanceTransferAllowedInCents = maxBalanceTransferAllowedInCents,
            Reward = reward,
            CardBackgroundHex = cardBackgroundHex
        };

        product.Apply(@event);
        return product;
    }

    public void Activate(string eventId, string correlationId)
    {
        if (IsActive)
            throw new InvalidOperationException("Product is already active");

        var @event = new ProductActivated
        {
            EventId = eventId,
            AggregateId = Id,
            AggregateVersion = Version + 1,
            CorrelationId = correlationId,
            CausationId = eventId,
            RecordedOn = DateTime.UtcNow
        };

        Apply(@event);
    }

    public void Deactivate(string eventId, string correlationId)
    {
        if (!IsActive)
            throw new InvalidOperationException("Product is already inactive");

        var @event = new ProductDeactivated
        {
            EventId = eventId,
            AggregateId = Id,
            AggregateVersion = Version + 1,
            CorrelationId = correlationId,
            CausationId = eventId,
            RecordedOn = DateTime.UtcNow
        };

        Apply(@event);
    }

    protected override void Apply(Event @event)
    {
        switch (@event)
        {
            case ProductDefined defined:
                Id = defined.AggregateId;
                Name = defined.Name;
                InterestInBasisPoints = defined.InterestInBasisPoints;
                AnnualFeeInCents = defined.AnnualFeeInCents;
                PaymentCycle = defined.PaymentCycle;
                CreditLimitInCents = defined.CreditLimitInCents;
                MaxBalanceTransferAllowedInCents = defined.MaxBalanceTransferAllowedInCents;
                Reward = defined.Reward;
                CardBackgroundHex = defined.CardBackgroundHex;
                IsActive = false;
                break;

            case ProductActivated:
                IsActive = true;
                break;

            case ProductDeactivated:
                IsActive = false;
                break;

            default:
                throw new ArgumentException($"Unknown event type: {@event.GetType().Name}");
        }
    }
}
